File: _scripts/mail/decompress/decompress-mails.py
# SCRIPT FOR DECOMPRESSING (ZSTD) MAILS IN MAILDIR FOLDER (ChatGPT)

import os
import zstandard as zstd
import tempfile
import shutil

# Pas dit aan naar jouw Maildir-map
MAILDIR = "./"

def is_zstd(file_path):
    """Controleer magic bytes van een zstd-bestand"""
    try:
        with open(file_path, "rb") as f:
            magic = f.read(4)
        return magic == b'\x28\xb5\x2f\xfd'
    except Exception:
        return False

def decompress_file_stream(file_path):
    """
    Streaming decompressie van een zstd-bestand.
    Het originele bestand wordt veilig overschreven via een tijdelijke file.
    """
    tmp_fd, tmp_path = tempfile.mkstemp(dir=os.path.dirname(file_path))
    os.close(tmp_fd)  # we gebruiken open() later

    try:
        with open(file_path, 'rb') as fin, open(tmp_path, 'wb') as fout:
            dctx = zstd.ZstdDecompressor()
            dctx.copy_stream(fin, fout)
        # Vervang origineel bestand door de decompressie
        shutil.move(tmp_path, file_path)
        print(f"[OK] Uitgepakt: {file_path}")
    except Exception as e:
        print(f"[FOUT] {file_path} kon niet uitgepakt worden: {e}")
        if os.path.exists(tmp_path):
            os.remove(tmp_path)

def process_maildir_cur(maildir_path):
    """
    Doorloop alle 'cur' submappen van de Maildir en decompress bestanden.
    """
    for root, dirs, files in os.walk(maildir_path):
        if os.path.basename(root) == 'cur':
            for filename in files:
                file_path = os.path.join(root, filename)
                if not is_zstd(file_path):
                    print(f"[SKIP] Geen zstd-bestand: {file_path}")
                    continue
                decompress_file_stream(file_path)

if __name__ == "__main__":
    process_maildir_cur(MAILDIR)

File: _scripts/mail/decompress/decompress-mails.php
<?php

$maildir = './';
$c=0;
$e=0;

/**
 * Controleer of een bestand een zstd-bestand is via magic bytes
 */
function is_zstd($filePath) : bool
{
    $fp = fopen($filePath, "rb");
    if (!$fp) return false;
    $magic = fread($fp, 4);
    fclose($fp);
    return $magic === "\x28\xB5\x2F\xFD";
}



/**
 * Functie die wordt uitgevoerd voor elk gevonden bestand.
 * Hier voer je je specifieke logica uit (bijv. lezen, schrijven, zstd decompressie, etc.).
 *
 * @param SplFileInfo $file Het bestandsobject met alle informatie over het bestand.
 */
/**
 * Decomprimeer een zstd-bestand en overschrijf het originele bestand
 */
function decompress_file($filePath) : void
{
    global $c,$e;
    $src = escapeshellarg($filePath);
    $tmp = $src . ".tmp";
    $command = "zstd -d " . $src . " -o " . $tmp;
    exec($command, $output, $returnVar);

    if ($returnVar === 0) {
        echo "[OK]";
        $c++;
        exec("mv -f {$tmp} {$src}", $output, $returnVar);
    } else {
        echo "[FOUT]: {$filePath}\n";
        $e++;
        exec("rm -f {$tmp}", $output, $returnVar);
    }
}

// --- De Recursieve Loop ---

try {
    // 1. Maak een RecursiveDirectoryIterator: dit gaat de directory structuur in.
    $iterator = new RecursiveDirectoryIterator($maildir);
    
    // 2. Maak een RecursiveIteratorIterator: dit zorgt ervoor dat de loop
    //    ALLE bestanden en submappen doorloopt.
    $files = new RecursiveIteratorIterator($iterator);
    
    echo "Start recursieve zoektocht in: {$maildir}\n";
    echo str_repeat('-', 40) . "\n";

    // Loop over elk item (bestand of map) dat de iterator vindt
    foreach ($files as $file) {
        
        if ($file->isDir()) { continue; }
        
        $f = $file->getPathname();
        
        // Controleer of het een ZSTD bestand is
        if (!is_zstd($f)) {
            echo "[SKIP] Geen zstd bestand: {$f}\n";
            continue;
        }
        
        decompress_file($f);
    }
    
    echo str_repeat('-', 40) . "\n";
    echo "   {$c} files successfully processed.\n";
    echo "   {$e} decompression errors encountered.\n\n";

} catch (UnexpectedValueException $e) {
    echo "[ERROR] Cannot open {$maildir}\n";
    echo "Message: " . $e->getMessage() . "\n";
}

?>

